### PR TITLE
Add amd64 bit to machine

### PR DESCRIPTION
**Pull request by rminnich**
This change allows callers of setup to specify
amd64.

This includes initial page tables and long
mode setup so we can start kernel at the
64-bit entry point.

It also includes support for KVM_TRANSLATE, which
is incredibly useful for debug.

golangci-lint is making a lot of bogus complaints.

Add a memsize flag, -m

usage
-m number -- default units are gigabytes

-m number[gGmMkK]

units of GiB, MiB, or KiB

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rminnich** [comment]
I've flagged all the golangci-lint anti-patterns and issues, because, I know that when I import this into Google again, the things golangci-lint requires will clash with google coding standards.

**Comment by rminnich** [comment]
I notice each CI involves a full kernel build. I'd suggest putting a binary of the kernel somewhere and skipping this step. It will make CI run in seconds, not minutes.

**Comment by rminnich** [comment]
I'm happy to host the kernel at github.com/linuxboot/bins if you want, we have many kernels there.

**Comment by rminnich** [comment]
WOOO HOOOOOO! 64-bit and ELF support now!

**Comment by rminnich** [comment]
I did resolve all the golangci-lint issues, though some of its complaints are a bit questionable :-)

That's ok, better to have it than not to have it :-)

**Review by bobuhiro11 — CHANGES_REQUESTED** [review]
Thanks for the great improvement of loading the ELF kernel.
It seems to work correctly.
I have written some comments inline for you to check.
```bash
ubuntu2004thinkpad:~/gokvm$ file _linux/vmlinux _linux/arch/x86/boot/bzImage
_linux/vmlinux:               ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=4acaf060bebd9131d2f051de66b4c5607473919a, with debug_info, not stripped
_linux/arch/x86/boot/bzImage: Linux kernel x86 boot executable bzImage, version 5.14.3 (bobuhiro11@ubuntu2004thinkpad) #1 SMP Wed Jan 18 02:11:50 UTC 2023, RO-rootFS, swap_dev 0x2, Normal VGA

ubuntu2004thinkpad:~/gokvm$ sudo go run . -k _linux/vmlinux
ubuntu2004thinkpad:~/gokvm$ sudo go run . -k _linux/arch/x86/boot/bzImage
```

**Comment by rminnich** [comment]
Thanks for the review!! I will work on this tomorrow.

File: go.mod [hunk 1]
```diff
@@ -1,3 +1,9 @@
 module github.com/bobuhiro11/gokvm
 
 go 1.17
+
+require (
+	golang.org/x/mod v0.7.0 // indirect
+	golang.org/x/sys v0.4.0 // indirect
+	golang.org/x/tools v0.5.0 // indirect
```
[the remark this answers is not in the document]
Comment (rminnich) [reply go.mod:8]: done

File: machine/machine.go [hunk 2]
```diff
@@ -20,45 +25,120 @@ import (
...
+//	              |                  |
+//	              +------------------+
+//	              |                  |
+//	0x0f000000    +------------------+ initrd [+ 0]
+//	              |                  |
+//	              |   initrd         |
+//	              |                  |
+//	              +------------------+
+//	              |                  |
+//	0x40000000    +------------------+
```
[the remark this answers is not in the document]
Comment (rminnich) [reply machine/machine.go:58]: done. I wonder why that happened?

File: machine/machine.go [hunk 3]
```diff
@@ -181,16 +263,56 @@ func New(kvmPath string, nCpus int, tapIfName string, diskPath string) (*Machine
...
+}
+
 // RunData returns the kvm.RunData for the VM.
 func (m *Machine) RunData() []*kvm.RunData {
 	return m.runs
 }
 
 func (m *Machine) LoadLinux(kernel, initrd io.ReaderAt, params string) error {
+	var (
+		DefaultKernelAddr = uint64(0x100000)
```
[the remark this answers is not in the document]
Comment (rminnich) [reply machine/machine.go:312]: done

File: machine/machine.go [hunk 4]
```diff
@@ -254,23 +390,52 @@ func (m *Machine) LoadLinux(kernel, initrd io.ReaderAt, params string) error {
...
+		amd64    bool
+		kernSize int
+	)
 
-	for i := range m.vcpuFds {
-		if err = m.initRegs(i); err != nil {
-			return err
+	switch isElfFile {
+	case false:
+		offset := int(bootParam.Hdr.SetupSects+1) * 512
```
[the remark this answers is not in the document]
Comment (rminnich) [reply machine/machine.go:401]: done

File: kvm/kvm.go [hunk 5]
```diff
@@ -193,23 +196,24 @@ func ioctl(fd, op, arg uintptr) (uintptr, error) {
...
 	Control  uint32
 	_        uint32
-	Debugreg [8]uint64
+	DebugReg [8]uint64
 }
 
-// SingleStep enables single stepping on on a vcpu.
-func SingleStep(vcpuFd uintptr, onoff bool) error {
+// SingleStep enables single stepping on all vCPUS in a VM. At present, it seems
+// not to work. It is based on working code from the linuxboot Voodoo project.
```
[the remark this answers is not in the document]
Comment (rminnich) [reply kvm/kvm.go:215]: I am generally worried about URLs in comments, and I wrote that code, and it's very basic, and that code was broken: I don't think we need that comment. I hope that is OK?

File: flag/flag.go [hunk 6]
```diff
@@ -2,28 +2,68 @@ package flag
...
+	flag.StringVar(&kernel, "k", "./bzImage", "kernel image path")
+	flag.StringVar(&initrd, "i", "./initrd", "initrd path")
 	//  refs: commit 1621292e73770aabbc146e72036de5e26f901e86 in kvmtool
-	params := flag.String("p", `console=ttyS0 earlyprintk=serial noapic noacpi notsc `+
+	flag.StringVar(&params, "p", `console=ttyS0 earlyprintk=serial noapic noacpi notsc `+
 		`debug apic=debug show_lapic=all mitigations=off lapic tsc_early_khz=2000 `+
 		`dyndbg="file arch/x86/kernel/smpboot.c +plf ; file drivers/net/virtio_net.c +plf" pci=realloc=off `+
 		`virtio_pci.force_legacy=1 rdinit=/init init=/init`, "kernel command-line parameters")
+	flag.StringVar(&tapIfName, "t", "tap", "name of tap interface")
+	flag.StringVar(&disk, "d", "/dev/zero", "path of disk file Var(&disk,for /dev/vda)")
```
[the remark this answers is not in the document]
Comment (rminnich) [reply flag/flag.go:52]: done

File: machine/machine.go [hunk 7]
```diff
@@ -201,10 +323,24 @@ func (m *Machine) LoadLinux(kernel, initrd io.ReaderAt, params string) error {
...
+	if err == nil {
+		isElfFile = true
+	}
+
+	bootParam := &bootparam.BootParam{}
+
+	// might be a bzimage
+	if !isElfFile {
+		// Load Boot Param
+		bootParam, err = bootparam.New(kernel)
```
[the remark this answers is not in the document]
Comment (bobuhiro11) [reply machine/machine.go:336]: If the bootparams are not present in the ELF, it seems fine to skip them.
Thanks!

File: kvm/kvm.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bobuhiro11) [reply kvm/kvm.go:215]: Sure. It's fine as is.

**Review by bobuhiro11 — APPROVED** [review]
Looks good. Thanks for your cooperation!

It took about 10 minutes to build the kernel. If there is no change in the kernel build config, it will be tuned so that it is just a simple download. I'll do that in another pull request.

**Pull request merged**